integration/singularity/store.go

```diff
@@ -457,3 +470,105 @@ func (s *SingularityStore) Describe(ctx context.Context, id blob.ID) (*blob.Desc
 	}
 	return descriptor, nil
 }
+
+func (s *SingularityStore) runCleanupWorker(ctx context.Context) {
+	s.closed.Add(1)
+	defer s.closed.Done()
+
+	// Run immediately once before starting ticker
+	s.runCleanupJob()
+
+	ticker := time.NewTicker(s.cleanupInterval)
+	for {
+		select {
+		case <-ticker.C:
+			s.runCleanupJob()
+		case <-s.closing:
+			return
+		}
+	}
+}
+
+func (s *SingularityStore) runCleanupJob() {
+	if err := s.cleanup(context.Background()); err != nil {
+		logger.Errorf("Local store cleanup failed: %w", err)
+	}
+}
+
+func (s *SingularityStore) cleanup(ctx context.Context) error {
+
+	logger.Infof("Starting local store cleanup...")
+
+	dir, err := os.ReadDir(s.local.Dir())
+	if err != nil {
+		return fmt.Errorf("failed to open local store directory: %w", err)
+	}
+
+	var binsToDelete []string
+
+binIteration:
+	for _, entry := range dir {
+		binFileName := entry.Name()
+
+		id, entryIsBin := strings.CutSuffix(binFileName, ".bin")
+		if !entryIsBin {
+			continue
+		}
+
+		idFileName := id + ".id"
+		idStream, err := os.Open(path.Join(s.local.Dir(), idFileName))
+		if err != nil {
+			logger.Warnf("Failed to open ID map file for %s: %v", id, err)
+			continue
+		}
+		fileIDString, err := io.ReadAll(idStream)
+		if err != nil {
+			logger.Warnf("Failed to read ID map file for %s: %v", id, err)
+			continue
+		}
+		fileID, err := strconv.ParseUint(string(fileIDString), 10, 64)
+		if err != nil {
+			logger.Warnf("Failed to parse file ID %s as integer: %v", fileIDString, err)
+			continue
+		}
+
+		getFileDealsRes, err := s.singularityClient.File.GetFileDeals(&file.GetFileDealsParams{
+			Context: ctx,
+			ID:      int64(fileID),
+		})
+		if err != nil {
+			logger.Warnf("Failed to get file deals for %v: %v", fileID, err)
+			continue
+		}
+
+		// Make sure the file has a deal for every SP
```

Comment:
I believe this outer-inner loop is inverted -- the SPs should be the outer, the deals the inner.

As it is, what we're doing is making sure every deal is for one of our SPs (so if we have 5 sps, and 1 deal for one of them, the loop as written will succeed). Instead, we should be determining if every SP has at least one deal.

Reply:
you're totally right, i can't believe i did that aaaahh. fixed!